doc/SAIP4Extensions/SAI_P4EXT_API_Proposal.md

```diff
@@ -0,0 +1,279 @@
+1. Ability to add exclusive features
+2. Expose device specific capabilities
+3. Rapid application prototyping
+4. Ability to introduce new features potentially without Recompiling SAI
+
+## Object Dependencies ##
+
+In the current SAI P4 Extension Proposal the newly introduced SAI_OBJECT_TYPE_P4EXT_ENTRY shall remain disconnected from the SAI Pipeline Object Graph. In other words there shall be no dependency from/to other SAI object to/from SAI_OBJECT_TYPE_P4EXT_ENTRY. Also the proposal places no restriction on the relative placement of the new features/table in the SAI Pipeline. The implementation and interaction of the new feature with the SAI pipeline is left to each individual vendor.Vendor A and Vendor B may choose to implement different sets of features using the new SAI P4 extension object. While a Vendor C may decide to not implement any new features using SAI P4 extensions.
+
+# API Specification #
+This section describes SAI P4 Extension API Proposal
+
+## New header file saip4ext.h ##
+
+The new header file defines interfaces for a single object of type P4EXT_ENTRY. This new object mimics a P4 table entity. The attributes of the object define the vairous P4 table constructs such as table name, match key, action name and action parameters. Each of these attributes is defined as  string type. For target that are P4 compatible these strings can simply be set to corresponding P4 table construct.
+
+### SAI P4EXT Entry Attributes ###
+*sai_p4ext_entry_attr_t* defines the SAI P4 Extension Attributes. Each of the attributes is of type string. As seen below the attributes mimic P4 table attributes such as table name, match fields (key:value pairs), action field (key:value pairs), action paramters (key:value pairs). The format for each of these string attribute values will be described later in the API usage section.
+
+```cpp
+/**
+ * @brief Attribute Id for P4 ext */
+typedef enum _sai_p4ext_entry_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_P4EXT_ENTRY_ATTR_START,
+
+    /**
+     * @brief SAI P4 EXT table id
+     *
+     * @type sai_s8_list_t
+     * @flags CREATE_AND_SET
+     * @default empty
+     */
+    SAI_P4EXT_ENTRY_ATTR_TABLE_ID = SAI_P4EXT_ENTRY_ATTR_START,
+
+    /**
+     * @brief SAI P4 EXT Match field
+     *
+     * @type sai_s8_list_t
+     * @flags CREATE_AND_SET
+     * @default empty
+     */
+    SAI_P4EXT_ENTRY_ATTR_MATCH_FIELD_ID,
```

Comment:
Do we need _ID in the name here?

Reply:
I agree that _ID suffix is not needed here and is misleading. I chose to retain the attr names from the original header file since they were already being used in PINS SONiC. But I can change them